### PR TITLE
Wip rgw bucket tests separation new

### PR DESCRIPTION
**Pull request by TRYTOBE8TME**
## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by TRYTOBE8TME** [comment]
Following are the links for the clean run of the tests in teuthology:

1. http://pulpito.front.sepia.ceph.com/kapandya-2021-01-28_18:53:49-rgw:notifications-master-distro-basic-smithi/
2. http://pulpito.front.sepia.ceph.com/kapandya-2021-01-28_16:43:37-rgw:multisite-master-distro-basic-smithi/

The 1st link is for the bucket_notifications tests and the 2nd is for the multisite tests.

**Comment by yuvalif** [comment]
> Following are the links for the clean run of the tests in teuthology:
> 
> 1. http://pulpito.front.sepia.ceph.com/kapandya-2021-01-28_18:53:49-rgw:notifications-master-distro-basic-smithi/
> 2. http://pulpito.front.sepia.ceph.com/kapandya-2021-01-28_16:43:37-rgw:multisite-master-distro-basic-smithi/
> 
> The 1st link is for the bucket_notifications tests and the 2nd is for the multisite tests.

failure in the multisite test (duplicate pubsub notifications where radosgw-admin iis used) was seen in other runs and not a regression here

**Comment by TRYTOBE8TME** [comment]
jenkins test make check

File: src/test/rgw/bucket_notification/README.rst [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+
+        ./bootstrap
+
+You will need to use the sample configuration file named ``bntests.conf.SAMPLE`` 
+that has been provided at ceph/src/test/rgw/bucket_notification. This file can be
+used to run the bucket notification tests on a Ceph cluster started with vstart.
+
+You also need to install Kafka which can be done by downloading and unzipping from the following::
+
+        https://www.apache.org/dyn/closer.cgi?path=/kafka/2.6.0/kafka-2.6.0-src.tgz
```
Comment (yuvalif) [inline src/test/rgw/bucket_notification/README.rst:20]: link is broken. please use: `https://archive.apache.org/dist/kafka/2.6.0/kafka-2.6.0-src.tgz`

File: src/test/rgw/bucket_notification/README.rst [hunk 2]
```diff
@@ -0,0 +1,43 @@
+============================
+ Bucket Notification tests
+============================
+
+The tests use the Nose test framework. To get started, ensure you have
+the ``virtualenv`` software installed; e.g. on Debian/Ubuntu::
+
+        sudo apt-get install python-virtualenv
+
+and then run::
```
Comment (yuvalif) [inline src/test/rgw/bucket_notification/README.rst:10]: where? in which directory?

File: src/test/rgw/bucket_notification/README.rst [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+
+        ./bootstrap
+
+You will need to use the sample configuration file named ``bntests.conf.SAMPLE`` 
+that has been provided at ceph/src/test/rgw/bucket_notification. This file can be
```
Comment (yuvalif) [inline src/test/rgw/bucket_notification/README.rst:15]: please put directory names as formatted text. e.g. `/directory/`
also, be consistent with `/path/to/ceph/src/...`

File: src/test/rgw/bucket_notification/README.rst [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+
+        KafkaClient {
+          org.apache.kafka.common.security.plain.PlainLoginModule required
+          username="alice"
+          password="alice-secret";
+        };
+
+After installing Kafka next is you need to start the Zookeeper and Kafka services.
+Here's the commands which can be used to start these services. For starting 
+Zookeeper service run::
```
Comment (yuvalif) [inline src/test/rgw/bucket_notification/README.rst:33]: i got this error:
```
Classpath is empty. Please build the project first e.g. by running './gradlew jar -PscalaVersion=2.13.2'
```

I think we should build before running the binaries

File: src/test/rgw/bucket_notification/README.rst [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+
+        bin/zookeeper-server-start.sh config/zookeeper.properties
+
+and then run to start the Kafka service::
+
+        bin/kafka-server-start.sh config/server.properties
```
Comment (yuvalif) [inline src/test/rgw/bucket_notification/README.rst:39]: please add the `-daemon` option, for running in background

File: src/test/rgw/bucket_notification/README.rst [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+          org.apache.kafka.common.security.plain.PlainLoginModule required
+          username="alice"
+          password="alice-secret";
+        };
+
+After installing Kafka next is you need to start the Zookeeper and Kafka services.
+Here's the commands which can be used to start these services. For starting 
+Zookeeper service run::
+
+        bin/zookeeper-server-start.sh config/zookeeper.properties
```
Comment (yuvalif) [inline src/test/rgw/bucket_notification/README.rst:35]: please add the -daemon option, for running in background

File: src/test/rgw/bucket_notification/README.rst [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+============================
+
+The tests use the Nose test framework. To get started, ensure you have
+the ``virtualenv`` software installed; e.g. on Debian/Ubuntu::
+
+        sudo apt-get install python-virtualenv
+
+and then run::
+
+        ./bootstrap
```
Comment (yuvalif) [inline src/test/rgw/bucket_notification/README.rst:12]: bootstrap required for sudo privileges.  why is this needed, given that we run in a virtualenv?

File: src/test/rgw/bucket_notification/README.rst [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+
+        bin/zookeeper-server-start.sh config/zookeeper.properties
+
+and then run to start the Kafka service::
+
+        bin/kafka-server-start.sh config/server.properties
+
+After starting vstart, zookeeper and kafka services you're ready to run the tests (when you're in /path/to/ceph/src/test/rgw/bucket_notification)::
+
+        BNTESTS_CONF=bntests.conf.SAMPLE ./virtualenv/bin/nosetests -s test_bn.py
```
Comment (yuvalif) [inline src/test/rgw/bucket_notification/README.rst:43]: can you please change the instructions so that we don't run the tests from inside the dev tree

**Review by yuvalif — CHANGES_REQUESTED** [review]
please see coments on the README file for running the tests locally

File: src/test/rgw/bucket_notification/README.rst (hunk 7 above)
Comment (TRYTOBE8TME) [reply src/test/rgw/bucket_notification/README.rst:12]: It would install the required packages

File: src/test/rgw/bucket_notification/README.rst (hunk 8 above)
Comment (yuvalif) [inline src/test/rgw/bucket_notification/README.rst:43]: i get this error:
```
-bash: ./virtualenv/bin/nosetests: No such file or directory
```

File: src/test/rgw/bucket_notification/README.rst (hunk 7 above)
Comment (yuvalif) [reply src/test/rgw/bucket_notification/README.rst:12]: if it installed them locally for the user or inside the virtualenv it DOES NOT NEED SUDO

File: src/test/rgw/bucket_notification/README.rst [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+
+You will need to use the sample configuration file named ``bntests.conf.SAMPLE`` 
+that has been provided at ``/path/to/ceph/src/test/rgw/bucket_notification``. This file can be
+used to run the bucket notification tests on a Ceph cluster started with vstart.
+
+You also need to install Kafka which can be done by downloading and unzipping from the following::
+
+        https://archive.apache.org/dist/kafka/2.6.0/kafka-2.6.0-src.tgz
+
+Then inside the kafka config directory (``/path/to/kafka-2.6.0-src/config``) you need to create a file named ``kafka_client_jaas.conf``
```
Comment (yuvalif) [inline src/test/rgw/bucket_notification/README.rst:22]:
```suggestion
Then inside the kafka config directory (``/path/to/kafka-2.6.0-src/config``) you need to create a file named ``kafka_server_jaas.conf``
```

**Comment by yuvalif** [comment]
jenkins test make check

File: src/test/rgw/bucket_notification/README.rst [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+
+After creating this above file run the following command in kafka directory (``/path/to/kafka-2.6.0-src/``)::
+
+        ./gradlew jar -PscalaVersion=2.13.2
+
+After following the above steps next is you need to start the Zookeeper and Kafka services.
+Here's the commands which can be used to start these services. For starting 
+Zookeeper service run::
+
+        bin/zookeeper-server-start.sh config/zookeeper.properties
```
Comment (yuvalif) [inline src/test/rgw/bucket_notification/README.rst:35]:
```suggestion
        bin/zookeeper-server-start.sh -daemon config/zookeeper.properties
```

File: src/test/rgw/bucket_notification/README.rst [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+
+and then run to start the Kafka service::
+
+        bin/kafka-server-start.sh config/server.properties
```
Comment (yuvalif) [inline src/test/rgw/bucket_notification/README.rst:39]:
```suggestion
        bin/kafka-server-start.sh -daemon config/server.properties
```

File: src/test/rgw/bucket_notification/README.rst [hunk 12]
```diff
@@ -0,0 +1,51 @@
...
+
+If you want to run Zookeeper and Kafka services in background add ``-daemon`` at the end of the command like::
+
+        bin/zookeeper-server-start.sh config/zookeeper.properties -daemon
```
Comment (yuvalif) [inline src/test/rgw/bucket_notification/README.rst:43]:
```suggestion
        bin/zookeeper-server-start.sh -daemon config/zookeeper.properties
```

File: src/test/rgw/bucket_notification/README.rst [hunk 13]
```diff
@@ -0,0 +1,51 @@
...
+
+and::
+
+        bin/kafka-server-start.sh config/server.properties -daemon
```
Comment (yuvalif) [inline src/test/rgw/bucket_notification/README.rst:47]:
```suggestion
        bin/kafka-server-start.sh -daemon config/server.properties 
```

File: src/test/rgw/bucket_notification/README.rst [hunk 14]
```diff
@@ -0,0 +1,51 @@
...
+
+After starting vstart, zookeeper and kafka services you're ready to run the tests (when you're in ``/path/to/ceph/src/test/rgw/bucket_notification/``)::
```
Comment (yuvalif) [inline src/test/rgw/bucket_notification/README.rst:49]: please do not run it from inside the source tree

File: src/test/rgw/bucket_notification/README.rst [hunk 15]
```diff
@@ -0,0 +1,51 @@
...
+
+        BNTESTS_CONF=bntests.conf.SAMPLE ./virtualenv/bin/python -m nose -s test_bn.py
```
Comment (yuvalif) [inline src/test/rgw/bucket_notification/README.rst:51]: if you want to use virtualenv we need instructions on it here.

File: src/test/rgw/bucket_notification/README.rst (hunk 14 above)
Comment (yuvalif) [inline src/test/rgw/bucket_notification/README.rst:49]:
```suggestion
After starting vstart, zookeeper and kafka services you're ready to run the tests::
```

File: src/test/rgw/bucket_notification/README.rst [hunk 16]
```diff
@@ -0,0 +1,51 @@
+============================
+ Bucket Notification tests
+============================
+
+The tests use the Nose test framework. To get started, ensure you have
+the ``virtualenv`` software installed; e.g. on Debian/Ubuntu::
+
+        sudo apt-get install python-virtualenv
+
+You will need to use the sample configuration file named ``bntests.conf.SAMPLE``
```
Comment (yuvalif) [inline src/test/rgw/bucket_notification/README.rst:10]:
```suggestion
You will need to use the sample configuration file named ``bntests.conf.SAMPLE`` . you can copy this file to the directory where you are running the tests and modify it if needed.
```

File: src/test/rgw/bucket_notification/README.rst (hunk 15 above)
Comment (yuvalif) [inline src/test/rgw/bucket_notification/README.rst:51]:
```suggestion
        BNTESTS_CONF=bntests.conf python -m nose -s /path/to/ceph/src/test/rgw/bucket_notification/test_bn.py
```

**Review by yuvalif — APPROVED** [review]

**Comment by yuvalif** [comment]
jenkins test make check

**Comment by yuvalif** [comment]
please wait for https://github.com/ceph/ceph/pull/39521 to merge, before merging this PR - so we won't get the failures fixed there.

File: qa/suites/rgw/notifications/frontend/civetweb.yaml [hunk 17]
```diff
@@ -0,0 +1,3 @@
+overrides:
+  rgw:
+    frontend: civetweb
```
Comment (cbodley) [inline qa/suites/rgw/notifications/frontend/civetweb.yaml:3]: you've made copies of several files that are meant to be symlinks. for example, under suites/rgw/verify/, frontend is a symlink to .qa/rgw_frontend (where .qa is symlink to ../.qa all the way up to /ceph/qa). this 'frontend' symlink lets us add or remove files in /ceph/qa/rgw_frontend and have those changes apply to every suite

the 'objectstore' and 'rgw_pool_type' symlinks works similarly - we shouldn't be copying any of those directly into this new suite

for a test suite that is specific to notifications, i don't really think it's valuable to test against multiple objectstores or pool types. that will just inflate the number of jobs in the suite without much benefit

File: qa/tasks/kafka.py [hunk 18]
```diff
@@ -0,0 +1,413 @@
+"""
+Deploy and configure Kafka for Teuthology
```
Comment (cbodley) [inline qa/tasks/kafka.py:2]: this kafka task seems to conflate two different things: setting up a kafka instance, and running the notification tests

it's probably worth splitting this into two separate tasks. having 'kafka' as a separate task means that we could potentially use the kafka instance for other things. having 'notification tests' as a separate task means that we could potentially run the tests against kafka, amqp, or whatever else we integrate

File: src/test/rgw/bucket_notification/test_bn.py [hunk 19]
```diff
@@ -0,0 +1,2822 @@
...
+    receiver.stop = True
+    task.join(1)
+    try:
+        receiver.consumer.close()
+    except Exception as error:
+        log.info('failed to gracefuly stop Kafka receiver: %s', str(error))
+
+
+def get_ip():
+    return 'localhost'
```
Comment (yuvalif) [inline src/test/rgw/bucket_notification/test_bn.py:410]: this cannot work with the HTTP tests.
RGW needs to send messages to the http server, however the http server is running on the test machine and not on the RGW machine, hence "localhost" is incorrect.
this is the root cause for the failure in [this teuthology run](http://qa-proxy.ceph.com/teuthology/kapandya-2021-02-25_07:17:03-rgw:notifications-master-distro-basic-smithi/5913170/)

Comment (yuvalif) [reply src/test/rgw/bucket_notification/test_bn.py:410]: the "get_ip()" function should be different in the case of http vs. kafka
in kafka - get_ip() should return localhost, as the kafka server is collocated with the RGW
while in case of http the http sever is collocated with the test script

File: qa/suites/rgw/notifications/frontend/civetweb.yaml (hunk 17 above)
Comment (yuvalif) [reply qa/suites/rgw/notifications/frontend/civetweb.yaml:3]: @cbodley as we discussed over the call. the frontend and the type of object pool, probably don't impact that test (the persistent queues are using the "log" pool).

Comment (cbodley) [reply qa/suites/rgw/notifications/frontend/civetweb.yaml:3]: in the standup, we agreed to run against a single configuration instead of supporting multiple frontends/pool types/etc

to generate the proper symlinks for this configuration, i ran these commands:
```
$ cd ceph/qa/suites/rgw/notifications
$ ln -s ../.qa .
$ ln -s .qa/rgw_frontend/beast.yaml .
$ ln -s .qa/objectstore/bluestore-bitmap.yaml .
$ ln -s .qa/clusters/fixed-2.yaml .
$ ln -s .qa/distros/supported-random-distro\$ .
$ git add .qa beast.yaml bluestore-bitmap.yaml fixed-2.yaml supported-random-distro\$
$ git rm -rf clusters frontend objectstore rgw_pool_type centos_latest.yaml
```
i pushed this branch with an extra commit to https://github.com/cbodley/ceph/commits/wip-39139-symlinks. feel free to cherry-pick that into the PR

Comment (TRYTOBE8TME) [reply qa/suites/rgw/notifications/frontend/civetweb.yaml:3]: Thank you so much, I'll cherry-pick the commit.

**Comment by TRYTOBE8TME** [comment]
Following are the links for the clean teuthology runs:

1) For bucket notification tests: http://pulpito.front.sepia.ceph.com/kapandya-2021-03-05_04:58:17-rgw:notifications-master-distro-basic-smithi/
2) For multisite tests: http://pulpito.front.sepia.ceph.com/kapandya-2021-03-05_05:41:54-rgw:multisite-master-distro-basic-smithi/

**Comment by yuvalif** [comment]
jenkins test make check

**Comment by yuvalif** [comment]
@TRYTOBE8TME the cause for the tests failure are python linter (flake8) warnings on your code:
```
./tasks/kafka.py:4:1: F401 'io.BytesIO' imported but unused
./tasks/kafka.py:5:1: F401 'io.StringIO' imported but unused
./tasks/kafka.py:6:1: F401 'configobj.ConfigObj' imported but unused
./tasks/kafka.py:7:1: F401 'base64' imported but unused
./tasks/kafka.py:8:1: F401 'argparse' imported but unused
./tasks/kafka.py:11:1: F401 'os' imported but unused
./tasks/kafka.py:12:1: F401 'random' imported but unused
./tasks/kafka.py:13:1: F401 'six' imported but unused
./tasks/kafka.py:14:1: F401 'string' imported but unused
./tasks/kafka.py:15:1: F401 'subprocess' imported but unused
./tasks/kafka.py:16:1: F401 'json' imported but unused
./tasks/kafka.py:17:1: F401 'sys' imported but unused
./tasks/kafka.py:18:1: F401 'pathlib.Path' imported but unused
./tasks/kafka.py:20:1: F401 'collections.OrderedDict' imported but unused
./tasks/kafka.py:21:1: F401 'itertools.chain' imported but unused
./tasks/kafka.py:25:1: F401 'teuthology.config.config as teuth_config' imported but unused
./tasks/kafka.py:27:1: F401 'teuthology.packaging.install_package' imported but unused
./tasks/kafka.py:28:1: F401 'teuthology.packaging.remove_package' imported but unused
./tasks/kafka.py:198:5: F841 local variable 'clients' is assigned to but never used
./tasks/notification_tests.py:17:1: F401 'teuthology.exceptions.ConfigError' imported but unused
```

File: qa/tasks/kafka.py [hunk 20]
```diff
@@ -70,6 +51,10 @@ def install_kafka(ctx, config):
                 args=['rm', '-rf', test_dir],
             )
 
+            ctx.cluster.only(client).run(
+                args=['rm', '-rf', '{tdir}/kafka-2.6.0-src.tgz'.format(tdir=teuthology.get_testdir(ctx))],
```
Comment (mattbenjamin) [inline qa/tasks/kafka.py:55]: it seems surprising to see a version hardcoded here

File: src/test/rgw/bucket_notification/test_bn.py (hunk 19 above)
Comment (mattbenjamin) [reply src/test/rgw/bucket_notification/test_bn.py:410]: did this get addressed?

**Comment by yuvalif** [comment]
jenkins test make check

File: qa/tasks/kafka.py (hunk 20 above)
Comment (TRYTOBE8TME) [reply qa/tasks/kafka.py]: Changed it and also working fine: http://pulpito.front.sepia.ceph.com/kapandya-2021-03-18_05:37:42-rgw:notifications-master-distro-basic-smithi/

**Comment by TRYTOBE8TME** [comment]
jenkins test make check

**Comment by cbodley** [comment]
thanks @TRYTOBE8TME, nice work
